File: Java/src/_0018_4Sum/Solution.java
package _0018_4Sum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 使用双索引技术解决
 * 时间复杂度：O(n*logn + n^3)
 * 空间复杂度：O(1)
 */
public class Solution {

    public List<List<Integer>> fourSum(int[] nums, int target) {

        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();

        if (n <= 3) return res;

        // 数组排序，默认升序排列
        Arrays.sort(nums);

        for (int i = 0; i <= n - 4; i = next_num_index(nums, i))
            for (int j = i + 1; j <= n - 3; j = next_num_index(nums, j)) {
                // 待查找的目标t
                int t = target - nums[i] - nums[j];
                if (nums[j + 1] + nums[j + 2] > t || nums[n - 1] + nums[n - 2] < t) continue;

                int p1 = j + 1, p2 = n - 1;
                while (p1 < p2) {
                    if (nums[p1] + nums[p2] == t) {
                        res.add(generateList(nums[i], nums[j], nums[p1], nums[p2]));
                        p1 = next_num_index(nums, p1);
                        p2 = pre_num_index(nums, p2);
                    } else if (nums[p1] + nums[p2] < t)
                        p1 = next_num_index(nums, p1);
                    else
                        p2 = pre_num_index(nums, p2);
                }
            }
        return res;
    }

    private List<Integer> generateList(int a, int b, int c, int d) {
        List<Integer> item = new ArrayList<>();
        item.add(a);
        item.add(b);
        item.add(c);
        item.add(d);
        return item;
    }

    // 返回nums[index+1...n)中第一个不等于nums[index]的数据的下标
    private int next_num_index(int[] nums, int index) {
        for (int i = index + 1; i < nums.length; i++)
            if (nums[i] != nums[index])
                return i;
        return nums.length;
    }

    // 返回nums[0...index-1]中第一个不等于nums[index]的数据的下标
    private int pre_num_index(int[] nums, int index) {
        for (int i = index - 1; i >= 0; i--)
            if (nums[i] != nums[index])
                return i;
        return -1;
    }

    public static void main(String[] args) {

        int[] nums = {1, 0, -1, 0, -2, 2};
        System.out.println((new Solution()).fourSum(nums, 0));
    }
}

File: Python/_0018_4Sum/Solution.py
# coding=utf-8
from typing import List

# 执行用时：1660 ms, 在所有 Python3 提交中击败了10.68%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了93.48%的用户
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()

        res = []
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, len(nums)):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                k = j + 1; u = len(nums) - 1
                while k < u:
                    if k > j + 1 and nums[k] == nums[k - 1]:
                        k += 1
                        continue
                    while k < u - 1 and nums[i] + nums[j] + nums[k] + nums[u - 1] >= target:
                        u -= 1
                    if nums[i] + nums[j] + nums[k] + nums[u] == target:
                        res.append([nums[i], nums[j], nums[k], nums[u]])
                    k += 1
        return res


if __name__ == "__main__":
    nums = [1, 0, -1, 0, -2, 2]
    print(Solution().fourSum(nums, 0))
